vm_phys.c

```diff
@@ -0,0 +1,211 @@
+
+/* Start of data section is start of pageable memory. */
+extern char __data_start[];
+/* End of .bss section is end of wired kernel memory. */
+extern char __ebss[];
+/* End of memory is marked as starting address of stack. */
+extern char _estack[]; 
+
+
+
+void vm_page_array_init(vm_paddr_t mem_start) {
+    size_t i = 0;
+    for(i = 0; i < NUM_OF_PAGES; i++) {
+        vm_page_array[i].phys_addr = mem_start + PAGE_SIZE * i;
+        vm_page_array[i].order = 0;
+    }
+}
+
+void vm_phys_print_free_blocks() {
+    int i = 0;
+    for(i = 0; i < VM_NFREEORDER; i++) {
+        kprintf("order %d:\n", i);
+        vm_page_t it;
+
+        TAILQ_FOREACH(it, &basic_seg.free_queues[i], freeq)
+            kprintf("(%p,%p) ", it->phys_addr, it->phys_addr+ORDER_SIZE(it->order) );
+        kprintf("\n");
+    }
+}
+
+void vm_phys_init() {
+
+    uintptr_t mem_start = (uintptr_t)__data_start;
+    uintptr_t mem_end = (uintptr_t)_estack;
+
+    if(ALIGN( mem_start, PAGE_SIZE) != (uintptr_t) mem_start)
+        FAIL("vm_phys_init: mem_start not page aligned ")
+
+    if(ALIGN( mem_end, PAGE_SIZE) != (uintptr_t) mem_end)
+        FAIL("vm_phys_init: mem_end not page aligned ")
+
+    if( (mem_end-mem_start)/PAGE_SIZE != NUM_OF_PAGES)
+        FAIL("vm_phys_init: Memory has wrong number of pages (%d), expected (%d)\n", 
+                (mem_end-mem_start)/PAGE_SIZE, 
+                NUM_OF_PAGES);
+
+    vm_page_array_init(mem_start);
+
+    /* Disallow allocating from kernel wired memory */
+    basic_seg.start = (vm_paddr_t) __ebss;
+    /* Leave some space for stack, and disallow allocating from it */
+    basic_seg.end = mem_end - STACK_SPACE*PAGE_SIZE;
+
+    basic_seg.start = ALIGN(basic_seg.start, PAGE_SIZE);
+    basic_seg.end = ALIGN(basic_seg.end, PAGE_SIZE);
+
+    int seg_size = (basic_seg.end - basic_seg.start) / PAGE_SIZE;
+
+    kprintf("physical memory allocator initiated: %d\n", seg_size);
+    kprintf("start of manageable memory: %p\n", basic_seg.start);
+    kprintf("end of manageable memory: %p\n", basic_seg.end);
+
+    int i;
+    for(i = 0; i < VM_NFREEORDER; i++)
+        TAILQ_INIT(&basic_seg.free_queues[i]);
+
+    /* Adding basic segments */
+    int ord = 0; /* order we're trying to find free list for */
+    int last_page = 0; /* index of last page not used */
+    for(ord = VM_NFREEORDER-1; ord >= 0; ord--)
+        while(ORDER_SIZE(ord) <= seg_size) {
+            /* Create free segment of order ord */
+            TAILQ_INSERT_TAIL(&basic_seg.free_queues[ord], &vm_page_array[last_page], freeq);
+            vm_page_array[last_page].order = ord;
+            seg_size -= ORDER_SIZE(ord);
+            last_page += ORDER_SIZE(ord);
+        }
+
+}
+
+/* Takes two pages which are buddies, and coalesce them */
+static vm_page_t coalesce(vm_page_t page1, vm_page_t page2) {
+    if(page1->order != page2->order)
+        FAIL("coalesce: trying to connect pages of different orders\n");
```

Comment:
You're the only user of `coalesce` function and you managed to put more defensive code into this function that actual functionality. I'd get rid of it.